### PR TITLE
Compute technology share as 'production / total production' (#288)

### LINKED ISSUE #288: target_market_share function generates wrong "projected" scenario
Hi,
I believe, function target_market_share generates wrong technology_share values for „projected” metric, when the following conditions are met:
- when only one company is in the matched loanbook, 
- flag by_company is set to TRUE 
- and weight_production is set to FALSE. 
In such situation, the technology_share for each technology is the same (for example, if in ALD dataset, a company has some capacity in 4 technologies within a given sector in year t, its capacity in year t+1 will be 25% for each technology, no matter what the starting share was in year t). 

Example code (based on documentation), where each out of 3 technologies has 0.33 share:

```r
your_loanbook <- loanbook_demo
your_ald <- ald_demo
matched <- match_name(your_loanbook, your_ald, min_score = 0.9) %>% prioritize()

#example company
matched <- matched[matched$name_direct_loantaker=='Toyota Motor Corporation',]

market_share_targets_company <- matched %>%
  target_market_share(
    ald = your_ald,
    scenario = scenario,
    by_company = TRUE,
    weight_production=FALSE
    
  )

tech_mix_data <- market_share_targets_company %>%
  filter(
    sector == "automotive",
    region == "global", 
    year == 2021,
    scenario_source =='etp_2017'
)


ggplot(
  data = tech_mix_data, 
  mapping = aes(
    x = metric, 
    y = technology_share,
    fill = technology
  )
) +
  geom_col(position='fill')+
  geom_text(aes(label=round(technology_share,2)), position="stack", size = ifelse(tech_mix_data$technology_share<0.02,0,7)) + 
  labs(x = "Metric", y = "Weighted Capacity [%]")

```

### PR DESCRIPTION
**Pull request by maurolepore**
Closes #288

This PR is WIP and includes a failing test. But I would like to discuss what seems to be a conflicting requirements. Passing the test that proves #288 is fixed produces the two failing tests below. @jdhoffa do you understand what's going on?  
                               
```r
Failure (test-target_market_share.R:223:3): with known input outputs as expected, at company level
out_target$production not equal to c(10, 90, 23.6, 152.9, 30, 95, 18, 57).
8/8 mismatches (average diff: 58.5)
[1] 1.00 -  10.0 ==   -9.0
[2] 1.00 -  90.0 ==  -89.0
[3] 1.68 -  23.6 ==  -21.9
[4] 1.68 - 152.9 == -151.2
[5] 1.00 -  30.0 ==  -29.0
[6] 1.00 -  95.0 ==  -94.0
[7] 0.60 -  18.0 ==  -17.4
[8] 0.60 -  57.0 ==  -56.4

Failure (test-target_market_share.R:368:3): outputs known value with `weight_production` (#131)
out_unweighted$projected$production not equal to 3.
1/1 mismatches
[1] 2 - 3 == -1
```

### CODE REVIEW

File: R/summarize_weighted_production.R [hunk 1]
```diff
@@ -72,9 +72,10 @@ summarize_unweighted_production <- function(data, ...) {
     )) %>%
     distinct() %>%
     group_by(.data$sector_ald, .data$technology, .data$year, ...) %>%
-    summarize(weighted_production = sum(.data$production), .groups = "keep") %>%
+    # FIXME: Confusing: `weighted_production` holds unweighted_production?
+    summarize(weighted_production = .data$production / sum(.data$production), .groups = "keep") %>%
     ungroup(.data$technology) %>%
-    mutate(weighted_technology_share = .data$weighted_production / sum(.data$weighted_production)) %>%
+    mutate(weighted_technology_share = .data$weighted_production) %>%
     group_by(!!!dplyr::groups(data))
```
Comment (maurolepore) [inline R/summarize_weighted_production.R:79]: I confess I got lost and managed to pass the test for #288 just by trying thing out, but along the way of fixing `technology_share` I seem to have broken `production`.

File: tests/testthat/test-target_market_share.R [hunk 2]
```diff
@@ -791,3 +791,69 @@ test_that("`technology_share` outputs consistently when multiple
...
+      region_isos = region,
+      by_company = TRUE,
+      weight_production = FALSE
+    ) %>%
+    filter(metric == "projected")
+
+  expect_equal(
+    out$technology_share,
+    out$production / sum(out$production)
+  )
```
Comment (maurolepore) [inline tests/testthat/test-target_market_share.R:858]: Here is the gist of the test that proves that #288 is fixed.

File: R/summarize_weighted_production.R [hunk 3]
```diff
@@ -80,9 +80,9 @@ summarize_unweighted_production <- function(data, ...) {
     distinct() %>%
     group_by(.data$sector_ald, .data$technology, .data$year, ...) %>%
     # FIXME: Confusing: `weighted_production` holds unweighted_production?
-    summarize(weighted_production = .data$production / sum(.data$production), .groups = "keep") %>%
-    ungroup(.data$technology) %>%
-    mutate(weighted_technology_share = .data$weighted_production) %>%
+    summarize(weighted_production = .data$production, .groups = "keep") %>%
+    ungroup(.data$technology, .data$tmsr, .data$smsp) %>%
```
Comment (jdhoffa) [inline R/summarize_weighted_production.R:84]: @maurolepore I reverted your changes as they did not have the desired effect. `weighted_production` should contain the actual `production` values (I agree the naming is annoying and confusing. it was chosen to ensure that the rest of the reformatting in the code downstream uses the correct column regardless of the value of `weight_production`). 

The solution I show here is simply ungrouping by `tmsr` and `smsp` (which are passed in as grouping arguments via `summary_groups`. This seems to solve the test issue, though I don't think it's the most elegant solution. I tried just removing `tmsr` and `smsp` from the argument `summary_groups`, but it seems they are in fact necessary for other parts of the calculation.

**Comment by jdhoffa** [comment]
When running @KapitanKombajn's original reprex, using `r2dii.analysis` built from this PR, I see the behaviour I expect: 
``` r
library(r2dii.data)
library(r2dii.match) 
library(r2dii.analysis)
library(dplyr, warn.conflicts = FALSE)
library(tidyverse)

#test datasets
your_loanbook <- loanbook_demo
your_ald <- ald_demo

matched <- match_name(loanbook_demo, ald_demo) %>%
  prioritize()

#select only one company to ease presentation
matched <- matched[matched$name_direct_loantaker=='Toyota Motor Corporation',]

#calculate results
market_share_targets_company<- matched %>%
  target_market_share(
    ald = ald_demo,
    scenario = scenario_demo_2020,
    region_isos = region_isos_demo,
    by_company = TRUE,
    weight_production = FALSE
    
  )

tech_mix_data <- market_share_targets_company %>%
  filter(
    sector == "automotive",
    region == "global", 
    metric == "projected",
    scenario_source =='demo_2020'
  )

#recalculation of technology_share
tmp<-aggregate(tech_mix_data$production, by=list(year=tech_mix_data$year), sum)
colnames(tmp)<-c("year", "total_production")

tech_mix_data_corrected<- merge(tech_mix_data,tmp)

tech_mix_data_corrected$technology_share_expected<- tech_mix_data_corrected$production/tech_mix_data_corrected$total_production

#graph of the original data (technology share is wrong starting from year 2021)
ggplot(
  data = tech_mix_data_corrected, 
  mapping = aes(
    x = year, 
    y = technology_share,
    fill = technology
  )
) +
  geom_col(position='fill')+
  geom_text(aes(label=round(technology_share,2)), position="stack") + 
  labs(x = "Production metric", y = "Technology Share [%]")
```



<sup>Created on 2021-05-04 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

**Review by jdhoffa — APPROVED** [review]

**Pull request merged**